### DISCUSSION TITLE
Offline dictionaries?

### DISCUSSION THREAD

**Discussion by jmthackett** [johnfactotum/foliate#1295] [Ideas]
Hey!

I love Foliate, it has displaced KOReader on my laptop as my main ebook reading software just because it is so beautiful and easy to use. So, first of all, thank you!

One of the things I did for KOReader was put together a stardict formatted english-esperanto dictionary which is here: https://github.com/jmthackett/en-eo-stardict/releases/tag/latest

I'm wondering, would you be open to offline dictionaries of this kind? (other formats are also good, naturally)

If so I'd be pretty happy to have a go at adding this. Just say the word :)

**Comment by johnfactotum** [comment]
Back in v2 it used `dictd` and `sdcv`, but this required sandbox escape, which is not ideal. And also `sdcv` doesn't really support HTML (if I recall correctly KORreader added this feature in their `sdcv` fork).

Now when developing v3, I wrote a pure JavaScript module that can read DICT and StarDict dictionaries: https://github.com/johnfactotum/foliate-js?tab=readme-ov-file#offline-dictionaries. The performance is extremely bad but should be OK for small to medium sized dictionaries. The advantage is that it would only need read-only permission to the directory storing the dictionaries. But I never had time to add this to Foliate so there's no offline dictionary support in v3. The main thing is that reading files into the WebView context is a bit involved (hacky actually, since it's technically kind of abusing the `run-file-chooser` signal).
